### PR TITLE
MAINT: Update constants to codata 2018 values (second try)

### PR DESCRIPTION
**Pull request by jakobjakobson13**
For issues see the comments on the commits.

### CODE REVIEW

File: scipy/constants/codata.py [hunk 1]
```diff
@@ -1184,7 +1551,12 @@ def parse_constants(d):
 for k in _physical_constants_2006:
     if 'momentum' in k:
         _aliases[k] = k.replace('momentum', 'mom.um')
-
+for k in _physical_constants_2018:
+    if 'momentum' in k:
+        _aliases[k] = k.replace('momentum', 'mom.um')
+    if 'vacuum mag. permeability' in k:
+        _aliases[k] = k.replace('vacuum mag. permeability', 'magn. constant')
```
Comment (jakobjakobson13) [inline scipy/constants/codata.py]: These line had did not generate any aliases when I checked them on my pc.

**Comment by jakobjakobson13** [comment]
My mistake, the replacement works fine. I'm however unsure about the last bit of code:
```
# finally, insert aliases for values
for k, v in list(_aliases.items()):
    if v in _current_constants:
        physical_constants[k] = physical_constants[v]
    else:
        print(k)
```

**Comment by rgommers** [comment]
> The replacement of the name with an alias does however not work as 'vacuum mag. permeability' is (already) part of the newest set of data and therefore it can't be replaced by its alias 'magn. constant'.

I don't fully follow this. Right now in master we have:
```
>>> from scipy import constants                                                
>>> constants.value('magn. constant')                                          
1.2566370614359173e-06
>>> constants.value('vacuum mag. permeability')                                
...
KeyError: 'vacuum mag. permeability'
```

So what is the issue exactly with giving `'magn. constant'` and `'vacuum mag. permeability'` both the same, updated value for this constant?

**Comment by jakobjakobson13** [comment]
> > The replacement of the name with an alias does however not work as 'vacuum mag. permeability' is (already) part of the newest set of data and therefore it can't be replaced by its alias 'magn. constant'.
> 
> I don't fully follow this. Right now in master we have:
> 
> ```
> >>> from scipy import constants                                                
> >>> constants.value('magn. constant')                                          
> 1.2566370614359173e-06
> >>> constants.value('vacuum mag. permeability')                                
> ...
> KeyError: 'vacuum mag. permeability'
> ```
> 
> So what is the issue exactly with giving `'magn. constant'` and `'vacuum mag. permeability'` both the same, updated value for this constant?

That is in principle not a big issue. It just won't work with the following code:

```python3
# finally, insert aliases for values
for k, v in list(_aliases.items()):
    if v in _current_constants:
        physical_constants[k] = physical_constants[v]
    else:
        del _aliases[k]
```

as it checks the names in the `physical_constants` against the names in the `_current_constants`. And with this code I don't see a way to set the alias `magn, constant` for the variable `vacuum mag. permeability`.

**Comment by jakobjakobson13** [comment]
The actual problem is not the parsing per se but the test functions:

```python3
=================================== FAILURES ===================================

__________________________________ test_find ___________________________________

[gw0] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python

    def test_find():

        keys = find('weak mixing', disp=False)

        assert_equal(keys, ['weak mixing angle'])

    

        keys = find('qwertyuiop', disp=False)

        assert_equal(keys, [])

    

        keys = find('natural unit', disp=False)

        assert_equal(keys, sorted(['natural unit of velocity',

                                    'natural unit of action',

                                    'natural unit of action in eV s',

                                    'natural unit of mass',

                                    'natural unit of energy',

                                    'natural unit of energy in MeV',

                                    'natural unit of mom.um',

                                    'natural unit of mom.um in MeV/c',

                                    'natural unit of length',

>                                   'natural unit of time']))

E       AssertionError: 

E       Items are not equal:

E       item=6

E       

E        ACTUAL: 'natural unit of momentum'

E        DESIRED: 'natural unit of mom.um'

keys       = ['natural unit of action', 'natural unit of action in eV s', 'natural unit of energy', 'natural unit of energy in MeV', 'natural unit of length', 'natural unit of mass', ...]

scipy/constants/tests/test_codata.py:25: AssertionError

______________________________ test_exact_values _______________________________

[gw2] linux -- Python 3.7.1 /home/travis/virtualenv/python3.7.1/bin/python

    def test_exact_values():

        # Check that updating stored values with exact ones worked.

        for key in codata.exact_values:

>           assert_((codata.exact_values[key][0] - value(key)) / value(key) == 0)

key        = 'magn. constant'

scipy/constants/tests/test_codata.py:52: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

scipy/constants/codata.py:1593: in value

    _check_obsolete(key)

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

key = 'magn. constant'

    def _check_obsolete(key):

        if key in _obsolete_constants and key not in _aliases:

            warnings.warn("Constant '%s' is not in current %s data set" % (

>               key, _current_codata), ConstantWarning)

E           scipy.constants.codata.ConstantWarning: Constant 'magn. constant' is not in current CODATA 2018 data set

key        = 'magn. constant'
```

**Comment by mdhaber** [comment]
Closes gh-10964.
@jakobjakobson13 if merging this will close issue 9590, please include text like above in a comment.

**Comment by jakobjakobson13** [comment]
Can someone help me out? Why are the tests after the last commit 253474b failing?

**Comment by rgommers** [comment]
Tests are failing because of
```
In [11]: _physical_constants_2002['atomic unit of mag. dipole moment']         
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-11-7a09a706b73b> in <module>
----> 1 _physical_constants_2002['atomic unit of mag. dipole moment']

KeyError: 'atomic unit of mag. dipole moment'

In [12]: _physical_constants_2002['atomic unit of magn. dipole moment']        
Out[12]: (1.8548019e-23, 'J T^-1', 1.6e-30)

In [13]: physical_constants['atomic unit of mag. dipole mom.']                 
Out[13]: (1.85480201566e-23, 'J T^-1', 5.6e-33)
```

Your changes to parsing have somehow caused the 2002 entry for `atomic unit of magn. dipole moment` to not be transformed to `atomic unit of mag. dipole mom.` anymore

**Comment by jakobjakobson13** [comment]
> Tests are failing because of
> 
> ```
> In [11]: _physical_constants_2002['atomic unit of mag. dipole moment']         
> ---------------------------------------------------------------------------
> KeyError                                  Traceback (most recent call last)
> <ipython-input-11-7a09a706b73b> in <module>
> ----> 1 _physical_constants_2002['atomic unit of mag. dipole moment']
> 
> KeyError: 'atomic unit of mag. dipole moment'
> 
> In [12]: _physical_constants_2002['atomic unit of magn. dipole moment']        
> Out[12]: (1.8548019e-23, 'J T^-1', 1.6e-30)
> 
> In [13]: physical_constants['atomic unit of mag. dipole mom.']                 
> Out[13]: (1.85480201566e-23, 'J T^-1', 5.6e-33)
> ```
> 
> Your changes to parsing have somehow caused the 2002 entry for `atomic unit of magn. dipole moment` to not be transformed to `atomic unit of mag. dipole mom.` anymore

Well, I renamed the name of the parsing function from `parse_constants` to `parse_constants_2002to2014` but I did not change the content of the function. And I also think there was no alias before my commits that changed `moment` to `mom.` in the 2002 values.

File: scipy/constants/codata.py [hunk 2]
```diff
@@ -1333,33 +1705,26 @@ def find(sub=None, disp=False):
...
+                'natural unit of energy in MeV',
+                'natural unit of mom.um',
+                'natural unit of mom.um in MeV/c',
+                'natural unit of length',
+                'natural unit of time']
 
 # finally, insert aliases for values
 for k, v in list(_aliases.items()):
-    if v in _current_constants:
+    if v in _current_constants or _tested_keys:
```
Comment (peterbell10) [inline scipy/constants/codata.py]: This is evaluated as `(v in _current_constants) or _tested_keys` which is always truthy. 

```suggestion
    if v in _current_constants or v in _tested_keys:
```

File: scipy/constants/codata.py [hunk 3]
```diff
@@ -1333,33 +1705,26 @@ def find(sub=None, disp=False):
...
-    val = _current_constants[key][0]
-    if abs(exact_values[key][0] - val) / val > 1e-9:
-        raise ValueError("Constants.codata: exact values too far off.")
-
-physical_constants.update(exact_values)
+exact_values = dict()
+for physical_constant in physical_constants:
+    if physical_constants[physical_constant][2] == 0.0:
+         exact_values[physical_constant] = physical_constants[physical_constant]
+physical_constants.update(exact_values)
```
Comment (peterbell10) [inline scipy/constants/codata.py]: Does this line do anything? `exact_values` were taken from `physical_constants` in the first place.

**Review by peterbell10 — DISMISSED** [review]

File: scipy/constants/codata.py (hunk 3 above)
Comment (jakobjakobson13) [reply scipy/constants/codata.py]: I'm not sure. I thought it wouldn't be very clever to fix the `constant values` to the selection 
```
exact_values = {
    'mag. constant': (mu0, 'N A^-2', 0.0),
    'electric constant': (epsilon0, 'F m^-1', 0.0),
    'characteristic impedance of vacuum': (sqrt(mu0 / epsilon0), 'ohm', 0.0),
    'atomic unit of permittivity': (4 * epsilon0 * pi, 'F m^-1', 0.0),
    'joule-kilogram relationship': (1 / (c * c), 'kg', 0.0),
    'kilogram-joule relationship': (c * c, 'J', 0.0),
    'hertz-inverse meter relationship': (1 / c, 'm^-1', 0.0)
}
```

that was hard coded prior to my commit. But I can reverse that if you want.

Comment (peterbell10) [reply scipy/constants/codata.py]: I was mainly commenting on the fact you're updating the dictionary with values from itself which should be equivalent to doing nothing.

However, now you mention it, is it possible that calculating the equations at double precision gives more precision than reading data from the table?

Comment (jakobjakobson13) [reply scipy/constants/codata.py]: > I was mainly commenting on the fact you're updating the dictionary with values from itself which should be equivalent to doing nothing.
Now as you point it out, that's true. 

> However, now you mention it, is it possible that calculating the equations at double precision gives more precision than reading data from the table?

I'm not sure Bout that one.

File: scipy/constants/codata.py (hunk 2 above)
Comment (jakobjakobson13) [reply scipy/constants/codata.py]: This was a step in the right direction but one issue remains: 
The alias substitution applies to the `physical_constants` whereas the `test_find` function searches the `_current_constants` that are unaffected by the alias mechanism. So how should we handle this issue?

Comment (jakobjakobson13) [reply scipy/constants/codata.py]: In this particular case, I'll tell what you tell me to do. I just don't want to be the guy that breaks backwards compatibility.

However, as cluttered this pull request already is, changing the aliases as well their replacement mechanism and adapting the test functions won't do any further harm.

File: scipy/constants/codata.py (hunk 3 above)
Comment (jakobjakobson13) [reply scipy/constants/codata.py]: Should be fixed.

File: scipy/constants/codata.py (hunk 2 above)
Comment (peterbell10) [reply scipy/constants/codata.py]: git blame shows that `find` was intentionally changed to use `_current_constants` by @pv in 1c274c25c08e31dc008dd698f2ab1f24b71c6cc0.

**Comment by jakobjakobson13** [comment]
So, in total you could say that this pull request got pretty messy. You can take it but if you want to solve the alias and test function issue another way, I won't blame you. The current situation is not perfect as the alias mechanism did not sufficiently handle the way how aliases were generated for the test function and so I had to change the test function.

So if you want to go with a more sophisticated approach, you have my blessing. If not, happy merging.

**Comment by jakobjakobson13** [comment]
Is there any news on your decision about this pull request? Can it be merged or do you want to refactor the codata value import mechanism in a more fundamental way?

**Comment by rgommers** [comment]
Both keys still work, so `find` returning a different key shouldn't really matter - that's not a real backwards compat break I'd say.
```
>>> constants.physical_constants['natural unit of momentum']                                   
(2.730924488e-22, 'kg m s^-1', 3.4e-30)
>>> constants.physical_constants['natural unit of mom.um']                                     
(2.730924488e-22, 'kg m s^-1', 3.4e-30)
```

Let's get this in. Thanks @jakobjakobson13!

**Pull request merged**